// @rstest/core@0.7.4/dist/0~62.js
import 'module';
/*#__PURE__*/ import.meta.url;
import { checkPkgInstalled } from "./913.js";
import { installGlobal, addDefaultErrorHandler } from "./0~681.js";
const environment = {
    name: 'jsdom',
    async setup (global, { jsdom = {} }) {
        checkPkgInstalled('jsdom');
        const { CookieJar, JSDOM, ResourceLoader, VirtualConsole } = await import("jsdom");
        const { html = '<!DOCTYPE html>', userAgent, url = 'http://localhost:3000', contentType = 'text/html', pretendToBeVisual = true, includeNodeLocations = false, runScripts = 'dangerously', resources, console = false, cookieJar = false, ...restOptions } = jsdom;
        const dom = new JSDOM(html, {
            pretendToBeVisual,
            resources: resources ?? (userAgent ? new ResourceLoader({
                userAgent
            }) : void 0),
            runScripts,
            url,
            virtualConsole: console && global.console ? new VirtualConsole().sendTo(global.console) : void 0,
            cookieJar: cookieJar ? new CookieJar() : void 0,
            includeNodeLocations,
            contentType,
            userAgent,
            ...restOptions
        });
        const cleanupGlobal = installGlobal(global, dom.window);
        const cleanupHandler = addDefaultErrorHandler(global);
        return {
            teardown () {
                cleanupHandler();
                dom.window.close();
                cleanupGlobal();
            }
        };
    }
};
export { environment };
