// @rstest/core/dist/0~809.js
import 'module';
/*#__PURE__*/ import.meta.url;
import { checkPkgInstalled } from "./913.js";
import { installGlobal, addDefaultErrorHandler } from "./0~681.js";
const environment = {
    name: 'happy-dom',
    async setup (global, { happyDom = {} }) {
        checkPkgInstalled('happy-dom');
        const { Window } = await import("happy-dom");
        const win = new Window({
            ...happyDom,
            url: happyDom.url || 'http://localhost:3000',
            console: console && global.console ? global.console : void 0
        });
        const cleanupGlobal = installGlobal(global, win, {
            additionalKeys: [
                'Request',
                'Response',
                'MessagePort',
                'fetch'
            ]
        });
        const cleanupHandler = addDefaultErrorHandler(global);
        return {
            async teardown () {
                cleanupHandler();
                if (win.close && win.happyDOM.abort) {
                    await win.happyDOM.abort();
                    win.close();
                } else await win.happyDOM.cancelAsync();
                cleanupGlobal();
            }
        };
    }
};
export { environment };
